### PR TITLE
sql: cast numeric values to floats by way of strings

### PR DESCRIPTION
**Pull request by sploiselle**
Should fix issues we saw with failing SQL logic tests; will move to this approach in `rust-dec`. @benesch You think this should be the approach in rust-dec, as well?

### CODE REVIEW

**Review by benesch — APPROVED** [review]
LGTM! I think it's fine if `rust-dec` wants to keep the other implementation, and maybe someone will eventually come along and figure out how to improve it directly. But I think it's also reasonable to switch it over to this. No strong opinion, I guess!

**Review by umanwizard — APPROVED** [review]
Haven't looked at the code, but my minimal SLT test case (that was failing before) passes with this change.

File: src/expr/src/scalar/func.rs [hunk 1]
```diff
@@ -367,18 +367,22 @@ fn cast_numeric_to_int64<'a>(a: Datum<'a>) -> Result<Datum<'a>, EvalError> {
 
 fn cast_numeric_to_float32<'a>(a: Datum<'a>) -> Result<Datum<'a>, EvalError> {
     let a = a.unwrap_numeric().0;
-    let mut cx = numeric::cx_datum();
-    match cx.try_into_f32(a) {
-        Ok(i) => Ok(Datum::from(i)),
-        Err(_) => Err(EvalError::Float32OutOfRange),
+    let i = a.to_string().parse::<f32>().unwrap();
+    if i.is_infinite() {
```
Comment (mjibson) [inline src/expr/src/scalar/func.rs:371]: Why is this needed? In postgres 14 you can do:

```
postgres=# select 'inf'::numeric::float4;
  float4  
----------
 Infinity
```

Comment (sploiselle) [reply src/expr/src/scalar/func.rs:371]: Oh, interesting--I didn't realize `'inf'::numeric` is valid as of PG 14. Our current numeric type aims for PG 13 semantics, so the code in question passes must in that regar––but I filed #7427 to track the work here.

I'm going to merge this as-is primarily because I'm OOO until next Tuesday, want to unbreak SLT, and will need a few days to implement PG 14-style infinity.

**Comment by philip-stoev** [comment]
This is https://github.com/MaterializeInc/materialize/issues/7394, so linked this PR and the ticket.
